### ISSUE TITLE
Export to ONNX with all tensor shapes included

### ISSUE THREAD

**Issue by gyenesvi** [pytorch/pytorch#52914]
## ❓ Questions and Help

Is it possible to export a PyTorch model to ONNX with all intermediate tensor shapes included so that shape inference need not be run when importing the ONNX model? Although the ONNX format seems to be able to represent such a case (in value_info fields), the function torch.onnx.export does not seem to have such an option. When a model is exported with ops not present in ONNX using torch.onnx.export(operator_export_type=ONNX_FALLTHROUGH), shape inference cannot be run on the resulting ONNX model.

If this is not possible now, it would be useful to include such an option in torch.onnx.export.

cc @houseroad @spandantiwari @lara-hdr @BowenBao @neginraoof @SplitInfinity

**Comment by garymm** [comment]
Specifically for the case where onnx.export can't convert an operator and `ONNX_FALLTHROUGH` kicks in, the shape inference will probably not be able to proceed through the un-converted operator, since it only operates on ONNX operators, not PyTorch operators.
Would it still be useful to you to have the intermediate shapes for the part of the graph before the unconvertible op?
If so, could you please explain the use case?
Thanks

**Comment by gyenesvi** [comment]
@garymm the goal would be exactly to avoid running shape inference on the ONNX model, because it will get stuck on un-converted operators. The assumption behind this is that in PyTorch before export, the shapes should be available for all tensors, since PyTorch can actually run the model (for example for a sample input), and those shapes calculated during the sample run could be saved into the ONNX model, hence avoiding the need to run shape inference later on the ONNX model itself. The point is to obtain all intermediate shapes, not just up to the first un-converted op.

The use case is simple: I want to export a model (given sample input shapes) into ONNX, so that I can optimize and further convert it to another format. The optimization / conversion step may require the shapes of some tensors to be known, which may be after the un-converted op. Optimization and further conversion can be done without understanding certain details of the un-converted op, such as the way its output shapes are calculated. It is enough if the runtime engine understands the details of the un-converted op, tools that optimize/convert the model don't need to if the unknown part is 'local' (i.e. does not effect other parts of the graph, for example via rendering all tensor shapes after a certain point unknown). Let me know if that explains it!

**Comment by garymm** [comment]
Thanks for explaining. This seems like a very uncommon use case so I don't think anyone on our team will work on it soon.

**Comment by gyenesvi** [comment]
@garymm can you explain what makes it uncommon? I believe a proprietary inference engine of most neural network accelerators in the edge domain will make optimization and conversion to their own IRs, at least implicitly when compiling an ONNX model for execution, so that does not seem like an uncommon thing to me. Or is it the fact that the model would have unknown operators that make it uncommon?
Currently, it seems to me that an ONNX model that has operators exported with fallthrough cannot be handled by any tool I know of (ONNX checker fails, ONNX shape inference does not go through, ONNX runtime cannot execute it, ONNX simplifier fails on it). So if that's right, what's the utility of such ONNX models? I believe having the actual tensor shapes in the ONNX model would at least somewhat remedy the situation and make it more easy to process such models. Furthermore, in industry applications, supporting custom operators also seems like a fairly important use case, since ONNX and its tools cannot be prepared to handle custom / proprietary / unpublished operators of companies using PyTorch / ONNX in their tool-chains.

**Comment by garymm** [comment]
Yes, it's the unknown operators that makes this uncommon. With just standard ONNX operators, I believe the desired information can be reconstructed with the ONNX shape inference APIs.

If I understand your use case correctly, there're at least two possible strategies here:

1. Register a custom symbolic function to have a particular ATen op converted to a custom ONNX op of your choosing. Shape inference on custom operators is supported by registering a symbolic function that outputs the custom op and specifies its output shapes and types. [For example](https://github.com/microsoft/onnxruntime/blob/963e1ace4ed21fd7e368a97b077fc474ff551022/onnxruntime/python/tools/pytorch_export_contrib_ops.py#L83-L84). **Would this work for your use case?**

1. Export with `ONNX_ATEN_FALLBACK` or `ONNX_FALLTHROUGH` and then try to handle the shape inference after the export. [This tool](https://github.com/microsoft/onnxruntime/blob/0292356bd7ce5b379be8ee9820bfbf4fd321ebc5/onnxruntime/python/tools/symbolic_shape_infer.py) seems to take that approach: it has special shape inference logic for a set of supported ATen ops, and it uses that to augment ONNX's built-in shape inference.

I guess the second approach would be easier with the shape information that you're asking for here, but it seems that tool is an existence proof that it's not strictly needed. If so, can you do something similar to what that tool is doing (or maybe even use that same tool)?

It's possible I'm misunderstanding something, in which case please point that out :-)

**Comment by gyenesvi** [comment]
Well your answers are going in the right direction, ones that I also considered but all of them got stuck somewhere.

> Shape inference on custom operators is supported by registering a symbolic function that outputs the custom op and specifies its output shapes and types.

That was my obvious first try, but did not seem to be solving the problem. I never came across a way of defining output shapes and types (possibly for multiple outputs) for custom ops, not to mention to doing so in a way that works across tools. For example, one Python tool would train a network and export in to ONNX, and another tool would read that ONNX, optimize and convert it. Could this scenario work out? Which tool would need to register the custom ops? Both? I guess the resulting ONNX model would not be self contained and transferrable to work in another Python environment. Do you have more detailed examples or documentation around this? The ones that I found did not cover shape inference, and as far as I understood they were only a means of _exporting_ custom ops to ONNX. But that part is also solved by exporting with fallthrough, so I did not find the registration to add any utility to that. Or does it have other utility besides export? Can other tools make use of that info, for example the model checker and the shape inference?

> it has special shape inference logic for a set of supported ATen ops, and it uses that to augment ONNX's built-in shape inference.

If there was a way of augmenting ONNX's built-in shape inference for custom ops, that could work. I only found a way to do so in C++, but not in Python. This would be part of a Python toolchain, to be run on various systems, so it would be really required to be done in Python without the need for compiling and installing C++ stuff.
I don't get how the referenced tool achieves that, to me it seems that all shape inference of known ops is redefined for some kind of symbolic case. Does this have some documentation? Can you explain what it does?

Thanks for the help.

**Comment by garymm** [comment]
The only thing resembling documentation that I can find is its unit tests:
https://github.com/microsoft/onnxruntime/blob/0292356bd7ce5b379be8ee9820bfbf4fd321ebc5/onnxruntime/test/python/onnxruntime_test_python_symbolic_shape_infer.py

It seems to take a model that has symbolic / dynamic shapes on inputs and produces a model that has intermediate shapes annotated with the appropriate symbolic shapes. That seems very similar to what you want, if I'm understanding correctly. Maybe you could try that tool out and see if you can understand what it does. Then send a PR to add some documentation, and perhaps another PR to enhance it to do what you want?

**Comment by gyenesvi** [comment]
Thanks, I'll see what I can get out of this, though honestly I am skeptical about it, since it does not seem to be augmenting ONNX's builtin shape inference, instead it seems to be rewriting the whole thing symbolically for known ops, and storing the calculated shapes in the model, which I could do just with the builtin shape inference as well (for fixed shapes), when there are no custom ops, but that's pointless. Furthermore, such tools have a fair chance of not being maintained in the future in parallel with ONNX itself (just as the whole ONNX runtime is lagging behind), so I would not build a company toolchain relying on this.

The custom op registration path seemed more promising; above you mentioned that custom shapes can also be defined, do you have some reference for doing that? To clarify, what I'd need is the ability to register a custom ops's shape inference function (it would probably have to be a python function I guess), and let the builtin shape inference run through the whole graph; then I could obtain all shapes.

**Comment by garymm** [comment]
The PyTorch exporter supports custom op registration with shape and type inference info via Python ([example](https://github.com/microsoft/onnxruntime/blob/963e1ace4ed21fd7e368a97b077fc474ff551022/onnxruntime/python/tools/pytorch_export_contrib_ops.py#L57)), but I believe ONNX shape inference does not.
That sounds like a reasonable feature request which you could file at https://github.com/onnx/onnx.

**Comment by gyenesvi** [comment]
It would be great to understand what that example actually does, as without any documentation (I haven't found any, only others searching for help about it), I can't really build on it. For example, if it's just for ONNX export, what are those shapes used for, since they are not stored in ONNX? Do they mean concrete shapes or some kind of symbolic ones? Moreover, is there a real advantage of registering the custom op for export, instead of just using fallthrough, besides having more control over mapping names/attributes? In the end, the ONNX model will just contain a custom op that all other tools cannot handle, right?

**Comment by thiagocrepaldi** [comment]
`torch.onnx.export` is in maintenance mode and we don't plan to add new operators/features or fix complex issues.

Please try the new ONNX exporter and reopen this issue with a full repro if it also doesn't work for you: [quick torch.onnx.dynamo_export API tutorial](https://pytorch.org/tutorials/beginner/onnx/export_simple_model_to_onnx_tutorial.html)
